optimism/src/mips/mod.rs

```diff
@@ -15,9 +15,176 @@
+    RType(SyscallExitGroup),
+    RType(SyscallReadHint),
+    RType(SyscallReadPreimage),
+    RType(SyscallReadOther),
+    RType(SyscallWriteHint),
+    RType(SyscallWritePreimage),
+    RType(SyscallWriteOther),
+    RType(SyscallFcntl),
+    RType(SyscallOther),
+    RType(MoveZero),
+    RType(MoveNonZero),
+    RType(Sync),
+    RType(MoveFromHi),
+    RType(MoveToHi),
+    RType(MoveFromLo),
+    RType(MoveToLo),
+    RType(Multiply),
+    RType(MultiplyUnsigned),
+    RType(Div),
+    RType(DivUnsigned),
+    RType(Add),
+    RType(AddUnsigned),
+    RType(Sub),
+    RType(SubUnsigned),
+    RType(And),
+    RType(Or),
+    RType(Xor),
+    RType(Nor),
+    RType(SetLessThan),
+    RType(SetLessThanUnsigned),
+    RType(MultiplyToRegister),
+    RType(CountLeadingOnes),
+    RType(CountLeadingZeros),
+    JType(Jump),
+    JType(JumpAndLink),
+    IType(BranchEq),
+    IType(BranchNeq),
+    IType(BranchLeqZero),
+    IType(BranchGtZero),
+    IType(BranchLtZero),
+    IType(BranchGeqZero),
+    IType(AddImmediate),
+    IType(AddImmediateUnsigned),
+    IType(SetLessThanImmediate),
+    IType(SetLessThanImmediateUnsigned),
+    IType(AndImmediate),
+    IType(OrImmediate),
+    IType(XorImmediate),
+    IType(LoadUpperImmediate),
+    IType(Load8),
+    IType(Load16),
+    IType(Load32),
+    IType(Load8Unsigned),
+    IType(Load16Unsigned),
+    IType(LoadWordLeft),
+    IType(LoadWordRight),
+    IType(Store8),
+    IType(Store16),
+    IType(Store32),
+    IType(Store32Conditional),
+    IType(StoreWordLeft),
+    IType(StoreWordRight),
+];
+
+impl<F: Field> CircuitTrait<MIPS_COLUMNS, Instruction, F, Env<F>> for MIPSCircuit<F> {
+    fn new(domain_size: usize, env: &mut Env<F>) -> Self {
+        let mut circuit = Self {
+            domain_size,
+            witness: HashMap::new(),
+            constraints: Default::default(),
+            lookups: Default::default(),
+        };
+
+        for instr in INSTRUCTIONS {
+            circuit.witness.insert(
+                instr,
+                Witness {
+                    cols: Box::new(std::array::from_fn(|_| Vec::with_capacity(domain_size))),
+                },
+            );
+            interpret_instruction(env, instr);
+            circuit.constraints.insert(instr, env.constraints.clone());
+            circuit.lookups.insert(instr, env.lookups.clone());
+            env.constraints = vec![]; // Clear the constraints for the next instruction
+            env.lookups = vec![]; // Clear the lookups for the next instruction
+        }
+        circuit
+    }
+
+    fn push_row(&mut self, instr: Instruction, row: &[F; MIPS_COLUMNS]) {
+        self.witness.entry(instr).and_modify(|wit| {
+            for (i, value) in row.iter().enumerate() {
+                if wit.cols[i].len() < wit.cols[i].capacity() {
+                    wit.cols[i].push(*value);
+                }
+            }
+        });
+    }
+
+    fn pad(&mut self, instr: Instruction) -> bool {
```

Comment:
A lot of these functions will be quite similar for many implementations... I'm wondering if it makes sense to abstract this.